src/demo.rs

```diff
@@ -1,59 +1,138 @@
-            "0f3fb1116e30963f1dc6631ad0cd7f00e324de7f3348264a1bba539fb4721c5d" => Some("020000000001010000000000000000000000000000000000000000000000000000000000000000ffffffff05023a030101ffffffff023e7e50090000000017a914b6dfeb91d03f56c1575d1b80c755db1ba029f2dd870000000000000000266a24aa21a9ed72b438543ba67fbbcfea524294609f96e2923969c4ef384e8660e8aa37544a240120000000000000000000000000000000000000000000000000000000000000000000000000"),
-            _ => None,
-        }.unwrap();
+impl<T> ElectrumBlockchain<T>
+where
+    T: Read + Write,
+{
+    pub fn new(electrum_client: Client<T>) -> Self {
+        Self::with_capacity(electrum_client, DEFAULT_CAPACITY)
+    }
 
-        let bytes: Vec<u8> = FromHex::from_hex(string).map_err(|_| ())?;
-        Ok(deserialize(&bytes).map_err(|_| ())?)
+    pub fn with_capacity(electrum_client: Client<T>, capacity: usize) -> Self {
+        ElectrumBlockchain {
+            electrum_client: RefCell::new(electrum_client),
+            utxo_set: RefCell::new(Vec::with_capacity(capacity)),
+            capacity,
+        }
     }
+}
 
-    fn is_unspent(&self, _txout: &OutPoint) -> Result<bool, Self::Error> {
-        Ok(true)
+impl<T> Blockchain for ElectrumBlockchain<T>
+where
+    T: Read + Write,
+{
+    type Error = super::Error;
+
+    fn get_tx(&self, txid: &Txid) -> Result<Transaction, Self::Error> {
+        self.electrum_client
+            .borrow_mut()
+            .transaction_get(txid)
+            .map_err(|x| x.into())
     }
 
-    fn get_random_utxo(&self) -> Result<OutPoint, Self::Error> {
-        Ok(OutPoint {
-            txid: Txid::from_hex(
-                "0f3fb1116e30963f1dc6631ad0cd7f00e324de7f3348264a1bba539fb4721c5d",
-            )
-            .unwrap(),
-            vout: 0,
-        })
+    fn is_unspent(&self, txout: &OutPoint) -> Result<bool, Self::Error> {
+        let script = &self.get_tx(&txout.txid)?.output[txout.vout as usize].script_pubkey;
+        let unspent_utxos = &self
+            .electrum_client
+            .borrow_mut()
+            .script_list_unspent(&script)?;
+        Ok(unspent_utxos.into_iter().any(|x| x.tx_hash == txout.txid))
     }
 
-    fn broadcast(&self, tx: &Transaction) -> Result<(), Self::Error> {
-        let bytes = serialize(tx);
-        debug!("Broadcasting: {}", bytes.to_hex());
-        Ok(())
+    fn get_random_utxo(
+        &self,
+        txout: &OutPoint,
+        seed: u64,
+    ) -> Result<Option<OutPoint>, Self::Error> {
+        let mut rng: StdRng = SeedableRng::seed_from_u64(seed);
+        if self.utxo_set.borrow().len() == 0 {
+            let mut txid = &txout.txid;
+            let mut tx = self.get_tx(&txid)?.clone();
+            let mut scripts = HashSet::new();
+
+            for _ in 0..rng.gen_range(1, 3) {
```

Comment:
I would do a few more than 3, maybe in the 10-20 range, idk

Reply:
IMHO we should consider how many steps you have to do, on average, before hitting a coinbase. Any idea how to calculate it?
My reason for putting (1, 3) was: I have less chances to hit a coinbase; but I understand this could leak privacy.